### ISSUE TITLE
[video_player] Add macOS support

### ISSUE THREAD

**Issue by JulianBissekkou** [flutter/flutter#41688]
I would like to play videos on macOS and windows which is currently not possible since the video_player plugin uses native views to do that. Is support for desktop planned or are there any plans to implement a flutter only video player?

**Comment by stuartmorgan** [comment]
Windows is already tracked in #37673; rescoping and retitling for the macOS part of this.

> since the video_player plugin uses native views to do that

It uses textures, not platform views, actually. This is blocked on #30717 which is underway.

**Comment by KevinTheGray** [comment]
@stuartmorgan, once #30717 is merged in, would the the next/final step be to build a `video_player_fde` plugin that supports macOS?   

It seems so, but just want to be sure I'm not missing any further requirements that I'm not seeing.

**Comment by stuartmorgan** [comment]
Yes (although it could potentially be submitted directly to flutter/plugins instead at this point; I need to add url_launcher as an initial experiment, but once the dust settles on that macOS support could start being added to other plugins directly).

There's an initial version of video_player for macOS linked from https://github.com/flutter/engine/pull/7087 that could be used as a starting point. I'm not sure offhand how much polish it still needs.

**Comment by cbenhagen** [comment]
@franciscojma86 what is the current status of this? Do you have some code already? Can I help?

**Comment by cbenhagen** [comment]
@stuartmorgan Are there good reasons against using platform views in the video player? Performance maybe? Or have they just not been around when the video_player was written? Early tests show that for DRM content we might not be able to use textures. @lesnitsky might be able to tell more about this. Also cc-ing @hterkelsen as he worked on platform views a lot lately.

**Comment by stuartmorgan** [comment]
> Do you have some code already?

See my previous comment for initial code.

> Are there good reasons against using platform views in the video player?

Well, for macOS that would make this blocked on platform view support. It would also make it inconsistent with the existing implementations.

Beyond that, I don't know if any pro/con analysis has been done on mobile. It definitely hasn't on desktop (since it's not possible yet).

> Or have they just not been around when the video_player was written?

I'm not familiar with the development of the mobile plugin.

**Comment by lesnitsky** [comment]
I'm working on a video player which supports HLS + Fairplay on macos and ios (AVFoundation) and DASH + Widevine on android (exo player). `copyPixelBuffer` of [AVPlayerItemVideoOutput](https://developer.apple.com/documentation/avfoundation/avplayeritemvideooutput/1386148-copypixelbuffer) always returns nil for DRM protected video, so I built a platform view powered videoplayer, HLS + Fairplay work well on ios, macos implementation is blocked by platform views support

**Comment by cbenhagen** [comment]
Sorry I was a bit unclear. I saw that code but wondered if someone is already working on more before I duplicate efforts. Also the ticket is assigned to someone so I am even more hesitant to take over.

As there is no texture support, `video_player_web` already uses platform views. So consistency is not really a concern. Supporting platform views in macOS is tracked in https://github.com/flutter/flutter/issues/41722.

@lesnitsky are you planning to integrate your work into the official `video_player` plugin?

**Comment by lesnitsky** [comment]
@cbenhagen if we can have a decision from flutter plugins maintainers that it's ok to replace existing implementation with platform views I will be happy to help. One more note: I'm using swift and kotlin, maybe that's a problem for flutter plugins repo. All plugins I've seen so far use Obj C/Java, there might be a good reason for it

**Comment by cbenhagen** [comment]
I think we should try to get more opinions on the pros and cons of using platform views.

If ever possible we should try to avoid segmentation. It makes it so much easier for everyone if there is only one, fully tested and widely used video_player. 

If I remember correctly, there once was a good reason not to use Swift. Apps had to ship their own Swift runtime which would increase the app size. As of Swift 5 this is not a problem anymore. `flutter_tool` defaults to Swift and Kotlin nowadays. 

I'd love to hear the opinion of @amirh on those questions.

**Comment by franciscojma86** [comment]
@cbenhagen I haven't been able to work on this. Feel free to take it. 

@lesnitsky macOS plugins are built on Swift so feel free to use it.

**Comment by cbenhagen** [comment]
I added a PR: https://github.com/flutter/plugins/pull/2437

**Comment by amirh** [comment]
The video player plugin was indeed implemented before we had platform views.
I think that using different embedding mechanisms on different platforms is unavoidable (e.g I know of custom embedder platforms that are using webview to play videos). We should favor the most efficient video embedding mechanism for each platform. E.g for Android the current VD based platform views mechanism adds costs over a texture based approach (higher min API requirement, performance).

**Comment by ThinkDigitalSoftware** [comment]
I don't know what I'm talking about, but why not integrate [vlckit](https://github.com/videolan/vlckit#features)?

**Comment by kanekv** [comment]
Looks like all dependencies are met, is this ready to proceed?

**Comment by stuartmorgan** [comment]
See discussion in the WIP PR referenced above; it's not clear that implementing this using `Texture`, rather than waiting for `PlatformView`, is a good investment of effort.

**Comment by cbenhagen** [comment]
Unless someone knows that [`PlatformView` ](https://github.com/flutter/flutter/issues/41722) will be available in the next few weeks, I would say it is worth porting the `Texture` based iOS plugin to macOS. I am currently working on a new port of the latest video_player code and filing new [issues](https://github.com/flutter/flutter/issues/78003) along the way. I have a version which runs on the bata and stable channels which are still using OpenGL but the recent switch to Metal crashes as soon as an external texture is used. I will try to create a good example of this and file another issue. If all of a sudden someone works on PlatformViews that would be greatly appreciated as it would also allow for DRM videos and more.

**Comment by cbenhagen** [comment]
https://github.com/flutter/flutter/issues/78056 includes a small test case for the `Texture` issue using Metal.

**Comment by cbenhagen** [comment]
I added a PR: https://github.com/flutter/plugins/pull/3712

**Comment by Kovalik98** [comment]
Hi! Currently there is an example of how you can play videos on macos?

**Comment by tuantiensiu** [comment]
We have any alternative solution for it? I'm doing develop app play video on desktop

**Comment by unoravi** [comment]
Any updates on this ??

**Comment by MeitanteiAshour** [comment]
Any updates?

**Comment by cbenhagen** [comment]
@MeitanteiAshour you can follow texture based PR at https://github.com/flutter/plugins/pull/3712

**Comment by Mohanedy98** [comment]
I wish that the flutter team work on solving this issue as soon as possible as  video player is an important feature for many apps that might avoid using flutter in desktop development because of missing this **"Basic Feature"**

**Comment by deimantasa** [comment]
Any updates on this one?

**Comment by stuartmorgan** [comment]
This is blocked on platform view support, per https://github.com/flutter/plugins/pull/3712#issuecomment-921025029

**Comment by deimantasa** [comment]
> This is blocked on platform view support, per [flutter/plugins#3712 (comment)](https://github.com/flutter/plugins/pull/3712#issuecomment-921025029)

Thank you for prompt reply, @stuartmorgan . If I understand correctly, there is sill no plan to make it work for `macos` due to this blockage? What'd you recommend using for folks, that are developing on `desktop`?

Mentioned `video_player_macos` does not seems to be maintained, which might be an issue. I've spotted that this guy (https://pub.dev/packages/dart_vlc) recently was updated, which might help. Any thoughts?

**Comment by stuartmorgan** [comment]
> If I understand correctly, there is sill no plan to make it work for `macos` due to this blockage?

The plan is to implement it using platform views, once platform view support is implemented.

> What'd you recommend using for folks, that are developing on `desktop`?

The Flutter team doesn't have any recommendations about specific third-party packages. It's an open ecosystem; we encourage people to evaluate options to see what fits their needs.

> Mentioned video_player_macos does not seems to be maintained

Per my comments in the PR, there's no reason that interested community members could not publish and maintain a `video_player_macos_texture` (or whatever other name they wanted to choose) based on the approach there.

**Comment by ryanaltair** [comment]
Any updates on this one?

**Comment by fzyzcjy** [comment]
I am having this issue when implementing https://github.com/fzyzcjy/flutter_screen_recorder - it is surprising to know that, the recording is even easier than playing...

**Comment by wang-bin** [comment]
[fvp](https://pub.dev/packages/fvp) is a video_player plugin supports all mobile and desktop platforms. Just adding 2 lines in your existing video_player code.

**Comment by ollyde** [comment]
@wang-bin nice, I'll give this a go

**Comment by sglim** [comment]
> @ryanaltair there is a package here https://pub.dev/packages/video_player_macos
> 
> It will be moved at some point when the Flutter team finally implements this.

It works perfectly. Thanks!

**Comment by stuartmorgan** [comment]
> This is blocked on platform view support, per [flutter/plugins#3712 (comment)](https://github.com/flutter/plugins/pull/3712#issuecomment-921025029)

Based on more recent conversations, in particular with the engine team, we now expect that we will need to support both texture-based and platform-view based video players for the foreseeable future; each has a different set of limitations. Given that we expect to need to continue to support the iOS texture-based version, and we have much better code-sharing mechanisms for iOS and macOS plugins, we've revisited the decision around having a texture-based version for macOS, and will be adding that.
